src/PDE_BPINN.jl

```diff
@@ -0,0 +1,502 @@
+        end
+
+        # note here no of samples referse to numensemble and points is the no of points in each dep_vars discretization
+        # each phi will give output in single domain of depvar(so we have each row as a vector of vector outputs)
+        # so we get after reduce a single matrix of n rows(samples), and j cols(points)
+        ensemblecurves = [Particles(reduce(vcat, preds[i])) for i in eachindex(phi)]
+
+        return ensemblecurves, estimatedLuxparams, estimated_params, timepoints
+    else
+        # get intervals for parameters corresponding to flux chains
+        Fluxparams = names[2]
+
+        # convert to format directly usable by Flux
+        estimatedFluxparams = [estimnnparams[Fluxparams[i]] for i in eachindex(phi)]
+
+        # infer predictions(preds) each row - NN, each col - ith sample
+        samplesn = reduce(hcat, samples)
+        preds = []
+        for j in eachindex(phi)
+            push!(preds,
+                [phi[j](timepoints[j], samplesn[:, i][Fluxparams[j]]) for i in 1:numensemble])
+        end
+
+        ensemblecurves = [Particles(reduce(vcat, preds[i])) for i in eachindex(phi)]
+
+        return ensemblecurves, estimatedFluxparams, estimated_params, timepoints
+    end
+end
+
+# priors: pdf for W,b + pdf for ODE params
+function ahmc_bayesian_pinn_pde(pde_system, discretization;
+        dataset = nothing, draw_samples = 1000,
+        bcstd = [0.01], l2std = [0.05],
+        phystd = [0.05], priorsNNw = (0.0, 2.0),
+        param = [], nchains = 1, Kernel = HMC,
+        Adaptorkwargs = (Adaptor = StanHMCAdaptor,
+            Metric = DiagEuclideanMetric, targetacceptancerate = 0.8),
+        Integratorkwargs = (Integrator = Leapfrog,),
+        MCMCkwargs = (n_leapfrog = 30,), saveats = [1 / 10.0],
+        numensemble = floor(Int, draw_samples / 3), progress = false, verbose = false)
+
+    pinnrep = symbolic_discretize(pde_system,
+        discretization,
+        bayesian = true,
+        dataset_given = dataset)
+
+    if discretization.param_estim && isempty(param)
+        throw(UndefVarError(:param))
+    elseif discretization.param_estim && dataset isa Nothing
+        throw(UndefVarError(:dataset))
+    elseif discretization.param_estim && length(l2std) != length(pinnrep.depvars)
+        throw(error("L2 stds length must match number of dependant variables"))
+    end
+
+    # for physics loglikelihood
+    full_weighted_loglikelihood = pinnrep.loss_functions.full_loss_function
+    chain = discretization.chain
+
+    if length(pinnrep.domains) != length(saveats)
+        throw(error("Number of independant variables must match saveat inference discretization steps"))
+    end
+
+    # NN solutions for loglikelihood which is used for L2lossdata
+    Φ = pinnrep.phi
+
+    # for new L2 loss
+    # discretization.additional_loss = 
+
+    if nchains > Threads.nthreads()
+        throw(error("number of chains is greater than available threads"))
```

Comment:
Why wouldn't that be allowed?

Reply:
sampling for each chain is done on a single thread, https://turinglang.org/AdvancedHMC.jl/stable/#Parallel-sampling

Reply:
but you could still have main chains than threads. Nothing stops that, it just batches